feature/visited/src/main/kotlin/uk/govuk/app/visited/VisitedViewModel.kt

```diff
@@ -60,10 +73,125 @@ internal class VisitedViewModel @Inject constructor(
+            text = "",
+            section = SCREEN_NAME,
+            action = SELECT_ALL_BUTTON
+        )
+    }
+
+    fun onDeselectAllClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = DESELECT_ALL_BUTTON
+        )
+    }
+
+    fun onDoneClick() {
+        analyticsClient.buttonFunction(
+            text = "",
+            section = SCREEN_NAME,
+            action = DONE_BUTTON
+        )
+    }
+
     fun onVisitedItemClicked(title: String, url: String) {
         viewModelScope.launch {
             visited.visitableItemClick(title = title, url = url)
         }
         analyticsClient.visitedItemClick(text = title, url = url)
     }
+
+    fun onRemove() {
+        _uiState.value?.visited?.forEach { (_, visitedItems) ->
+            visitedItems.forEach { visitedItem ->
+                if (visitedItem.isSelected) {
+                    viewModelScope.launch {
+                        visitedLocalDataSource.remove(visitedItem.title, visitedItem.url)
+                        onRemoveVisitedItem(visitedItem.title)
+                    }
+                }
+            }
+        }
+    }
+
+    fun onSelectAll() {
+        _uiState.value = _uiState.value?.copy(
+            visited = _uiState.value?.visited?.mapValues { (_, visitedItems) ->
+                visitedItems.map { it.copy(isSelected = true) }
+            },
+            hasSelectedItems = true,
+            hasAllSelectedItems = true
+        )
+    }
+
+    fun onDeselectAll() {
+        _uiState.value = _uiState.value?.copy(
+            visited = _uiState.value?.visited?.mapValues { (_, visitedItems) ->
+                visitedItems.map { it.copy(isSelected = false) }
+            },
+            hasSelectedItems = false,
+            hasAllSelectedItems = false
+        )
+    }
+
+    fun onSelect(title: String, url: String) {
+        _uiState.value = _uiState.value?.let { currentState ->
+            val updatedVisited = currentState.visited?.mapValues { (_, visitedItems) ->
+                visitedItems.map {
+                    if (it.title == title && it.url == url) {
+                        it.copy(isSelected = !it.isSelected)
+                    } else {
+                        it
+                    }
+                }
+            }
+
+            val hasAllSelectedItems = updatedVisited?.all { (_, visitedItems) ->
+                visitedItems.all { it.isSelected }
+            } ?: false
+
+            val hasSelectedItems = updatedVisited?.any { (_, visitedItems) ->
+                visitedItems.any { it.isSelected }
+            } ?: false
+
+            currentState.copy(
+                visited = updatedVisited,
+                hasAllSelectedItems = hasAllSelectedItems,
+                hasSelectedItems = hasSelectedItems
+            )
+        }
+    }
+
+    internal fun getUiStateForTest(): VisitedUiState? = _uiState.value
```

Comment:
Apologies if I've misread but can we not just read public `uiState` here instead of this function?

Reply:
I don't really like adding test specific functions

Reply:
Fixed up